offchainservice/src/main/java/pl/edu/agh/blockchain/offchainservice/service/RegistrationService.java

```diff
@@ -0,0 +1,113 @@
+import pl.edu.agh.blockchain.offchainservice.repository.EmailRepository;
+import pl.edu.agh.blockchain.offchainservice.repository.PendingRepository;
+import pl.edu.agh.blockchain.offchainservice.utils.CodeGenerator;
+
+import java.io.IOException;
+import java.text.MessageFormat;
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class RegistrationService {
+
+    private final EmailRepository emailRepository;
+
+    private final PendingRepository pendingRepository;
+
+    private final EMailService eMailService;
+
+    private final DeploymentService deploymentService;
+
+    private final ModelMapper modelMapper = new ModelMapper();
+
+    public RegistrationService(EmailRepository emailRepository, EMailService eMailService,
+                               PendingRepository pendingRepository, DeploymentService deploymentService) {
+        this.emailRepository = emailRepository;
+        this.eMailService = eMailService;
+        this.pendingRepository = pendingRepository;
+        this.deploymentService = deploymentService;
+    }
+
+    public void verifyUser(String email) {
+        if (!email.endsWith("@agh.edu.pl") && !email.endsWith("@student.agh.edu.pl")) {
+            throw new InvalidEmailDomainException(email + " is not AGH email");
+        }
+        if (!emailRepository.existsByEmail(email)) {
+            try {
+                String token = CodeGenerator.generateNewToken();
+                eMailService.sendEmail(email, token);
+                savePendingVerification(email, token);
+            } catch (IOException e) {
+                throw new SendFailedException(e.getMessage());
+            }
+        } else {
+            User user = emailRepository.findByEmail(email).get(0);
+            throw new UserAlreadyExistsException(MessageFormat.format("User with mail: {0} was registered: {1}", user.getEmail(),
+                    user.getCreationDate()));
+        }
+    }
+
+    public void registerUser(RegisterInformationDTO registerInformationDTO) {
+        Optional<PendingVerificationDTO> pendingVerificationDTO = registrationPending(registerInformationDTO.getEmail());
+        if (pendingVerificationDTO.isPresent()) {
+            PendingVerificationDTO pendingVerification = pendingVerificationDTO.get();
+            if (!correctToken(registerInformationDTO, pendingVerification)) {
+                throw new InvalidVerificationToken("Token is not correct. Check your mail once again or wait about 20 minutes and generate new token");
+            }
+            saveUser(registerInformationDTO.getEmail());
+            removeFromPending(registerInformationDTO.getEmail());
+            deploymentService.deployContract(registerInformationDTO.getBlockchainAddress());
+        } else {
+            throw new VerificationTimeException("It's been 15 minutes since the email was sent. Please generate the verification email again");
```

Comment:
This is the case when the token expired (the user verified over 15 mins ago) or the user never verified. Reflect those two possibilities in the error message

Reply:
I think he shouldn't have option to verify himself before request for email with token.  He can go to the verify window, only when pass email. If he never verified then email will be sent. If he already verified, he will get information and app go to next window